### PR TITLE
AFNO solar radiation diagnostic model

### PR DESCRIPTION
**Pull request by loliverhennigh**
# Earth2Studio Pull Request

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.

### CODE REVIEW

File: earth2studio/models/dx/solarradiation_afno.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+            cache_options={
+                "cache_storage": Package.default_cache("ssrd_afno"),
+                "same_names": True,
+            },
+        )
+        return package
+
+    @classmethod
+    @check_extra_imports("solarradiation-afno", [SolarRadiationNet, cos_zenith_angle])
+    def load_model(cls, package: Package, freq: str = "6h") -> DiagnosticModel:
```
Comment (NickGeneva) [inline earth2studio/models/dx/solarradiation_afno.py]: freq: str = "6h"

I know we discussed but jotting this down for tracking
Can either make this two classes or make this an enum

File: earth2studio/models/dx/solarradiation_afno.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+    "q850",
+    "q925",
+    "q1000",
+]
+
+
+@check_extra_imports("solarradiation-afno", [SolarRadiationNet, cos_zenith_angle])
+class SolarRadiationAFNO(torch.nn.Module, AutoModelMixin):
+    """Solar Radiation AFNO diagnostic model. Predicts the accumulated global surface solar
+    radiation over 6 hours [Jm^-2]. The model uses 31 variables as input and outputs
```
Comment (NickGeneva) [inline earth2studio/models/dx/solarradiation_afno.py]: todo: update the 6 hours doc string

File: earth2studio/models/dx/solarradiation_afno.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+        return output_coords
+
+    def __str__(self) -> str:
+        return "SolarRadiationNet"
+
+    @classmethod
+    def load_default_package(cls) -> Package:
+        """Load prognostic package"""
+        package = Package(
+            "ngc://models/nvidian/onboarding/afno_dx_solarradiation@1",
```
Comment (NickGeneva) [inline earth2studio/models/dx/solarradiation_afno.py]: Update:

`ngc://models/nvidian/onboarding/afno_dx_solarradiation@1` -> `ngc://models/nvidia/earth-2/afno_dx_solarradiation@1.0.0`

File: earth2studio/models/dx/solarradiation_afno.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+                "cache_storage": Package.default_cache("ssrd_afno"),
+                "same_names": True,
+            },
+        )
+        return package
+
+    @classmethod
+    @check_extra_imports("solarradiation-afno", [SolarRadiationNet, cos_zenith_angle])
+    def load_model(cls, package: Package, freq: str = "6h") -> DiagnosticModel:
+        """Load diagnostic from package"""
```
Comment (NickGeneva) [inline earth2studio/models/dx/solarradiation_afno.py]: Improve this doc string to:

```
"""Load prognostic from package

        Parameters
        ----------
        package : Package
            Package to load model from

        Returns
        -------
        PrognosticModel
            Prognostic model
        """
  ```

File: earth2studio/models/dx/solarradiation_afno.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+        if SolarRadiationNet is None or cos_zenith_angle is None:
```
Comment (NickGeneva) [inline earth2studio/models/dx/solarradiation_afno.py]: Check can be removed now, should be handled by decorator

File: earth2studio/models/dx/solarradiation_afno.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+        output_coords : CoordSystem
+            Output coordinate system
+
+        Returns
+        -------
+        torch.Tensor
+            Solar zenith angle tensor
+        """
+        lon, lat = self.get_sza_lonlat(output_coords["lon"], output_coords["lat"])
+        t = output_coords["time"] + output_coords["lead_time"]
```
Comment (NickGeneva) [inline earth2studio/models/dx/solarradiation_afno.py:283]: Will this create a 2D matrix?

For example if time has shape 2 and lead_time has shape 2 there are 4 unique times we need the sza for.

I think this needs something like:

```
>>> from datetime import datetime, timedelta
>>> coords = {"time": np.array([datetime(2025,1,1), datetime(2024,8,1)]), "lead_time": np.array([timedelta(hours=6), timedelta(hours=12)])}
>>> t = coords['time'][:,None] + coords['lead_time'][None,:]
>>> t= t.flatten()
>>> t
array([datetime.datetime(2025, 1, 1, 6, 0),
       datetime.datetime(2025, 1, 1, 12, 0),
       datetime.datetime(2024, 8, 1, 6, 0),
       datetime.datetime(2024, 8, 1, 12, 0)], dtype=object)
```

File: earth2studio/models/dx/solarradiation_afno.py [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+        x = x.reshape(
+            -1, *x.shape[3:]
+        )  # Combine batch, time, lead_time into one dimension
+
+        # Normalize input
+        x = (x - self.era5_mean) / self.era5_std
+        output_coords = self.output_coords(coords)
+
+        # compute solar zenith angle and concatenate
+        sza = self.compute_sza(output_coords).reshape((1, 1, *x.shape[2:])).to(x.device)
```
Comment (NickGeneva) [inline earth2studio/models/dx/solarradiation_afno.py]: With `reshape((1, 1, *x.shape[2:]))`

Doesnt this assume a time dim of size 1, should then be `reshape((1, *x.shape[1:]))` with the edit suggested above?

Comment (NickGeneva) [reply earth2studio/models/dx/solarradiation_afno.py]: Oh x is already collapsed here, but im still not following the dims of what the second 1 is, is that for variable? If so then shouldnt the first dim be time*lead_time?

Perhaps its easier to reshape to `.reshape((1, time_size, lead_time_size, *x.shape[3:]))`
Then repeat dim0 for batch_size, then reshape like x above.

File: test/models/dx/test_solarradiation_afno.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        ssrd_std=ssrd_std,
+        orography=orography,
+        landsea_mask=landsea_mask,
+        sincos_latlon=sincos_latlon,
+    ).to(device)
+    x = x.to(device)
+    coords = OrderedDict(
+        {
+            "batch": np.ones(x.shape[0]),
+            "time": np.array([np.datetime64("2024-01-01")]),
```
Comment (NickGeneva) [inline test/models/dx/test_solarradiation_afno.py]: In the test parameterizations please add a test for multiple times single lead time, single time multiple lead time and then multiple time and lead time to check the model behaves correctly

File: earth2studio/models/dx/solarradiation_afno.py [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+
+        Returns
+        -------
+        CoordSystem
+            Coordinate system dictionary
+        """
+        target_input_coords = self.input_coords()
+        handshake_dim(input_coords, "lon", 5)
+        handshake_dim(input_coords, "lat", 4)
+        handshake_dim(input_coords, "variable", 3)
```
Comment (NickGeneva) [inline earth2studio/models/dx/solarradiation_afno.py:158]: should also handshake lead_time 2 and time 1

File: earth2studio/models/dx/solarradiation_afno.py [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+        x = torch.cat(
+            (x, sza, repeat_sincos_latlon, repeat_orography, repeat_landsea_mask), dim=1
+        )
+
+        repeat_ssrd_mean = self.ssrd_mean.repeat(x.shape[0], 1, 1, 1)
+        repeat_ssrd_std = self.ssrd_std.repeat(x.shape[0], 1, 1, 1)
+        out = self.core_model(x) * repeat_ssrd_std + repeat_ssrd_mean
+
+        # filter out negative values
+        out[out < 0] = 0
```
Comment (NickGeneva) [inline earth2studio/models/dx/solarradiation_afno.py]: out = torch.clamp(out, min=0) instead? Probably faster than condition index I would guess

File: earth2studio/models/nn/afno_ssrd.py [hunk 11]
```diff
@@ -0,0 +1,322 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023 - 2024 NVIDIA CORPORATION & AFFILIATES.
```
Comment (NickGeneva) [inline earth2studio/models/nn/afno_ssrd.py]: This isnt the right header, will through an error

File: earth2studio/models/dx/solarradiation_afno.py [hunk 12]
```diff
@@ -0,0 +1,363 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
```
Comment (NickGeneva) [inline earth2studio/models/dx/solarradiation_afno.py]: Check license header

**Review by NickGeneva — CHANGES_REQUESTED** [review]
Mostly just minor nits, main concern is the bug I think is present with the zenith angle calc

**Comment by loliverhennigh** [comment]
Ill try to find some comparison of this tomorrow but looking at other plots on the internet of the solar radiation this is about the right range.

02_solar_radiation_prediction

```
import os

os.makedirs("outputs", exist_ok=True)
from dotenv import load_dotenv

load_dotenv()  # TODO: make common example prep function

from earth2studio.data import GFS
from earth2studio.io import ZarrBackend
from earth2studio.models.dx import SolarRadiationAFNO6H
from earth2studio.models.px import SFNO

# Load the default model package which downloads the check point from NGC
package = SFNO.load_default_package()
prognostic_model = SFNO.load_model(package)

package = SolarRadiationAFNO6H.load_default_package()
diagnostic_model = SolarRadiationAFNO6H.load_model(package)

# Create the data source
data = GFS()

# Create the IO handler, store in memory
io = ZarrBackend()

import earth2studio.run as run

nsteps = 2
io = run.diagnostic(
    ["2021-06-01"], nsteps, prognostic_model, diagnostic_model, data, io
)

print(io.root.tree())

from datetime import datetime

import cartopy.crs as ccrs
import matplotlib.pyplot as plt
import numpy as np

forecast = datetime(2021, 6, 1)
variable = "ssrd"  # Surface solar radiation
step = 2  # lead time = 48 hrs

plt.close("all")
# Create a Orthographic projection of USA
projection = ccrs.Orthographic(-100, 40)

# Create a figure and axes with the specified projection
fig, ax = plt.subplots(subplot_kw={"projection": projection}, figsize=(10, 6))

# Plot the field using pcolormesh
levels = np.linspace(0, 1e6, 11)  # Range for solar radiation in J/m^2
im = ax.contourf(
    io["lon"][:],
    io["lat"][:],
    io[variable][0, step],
    levels,
    transform=ccrs.PlateCarree(),
    cmap="YlOrRd",
)

# Set title
ax.set_title(f"{forecast.strftime('%Y-%m-%d')} - Lead time: {6*step}hrs")

# Add coastlines and gridlines
ax.set_extent([220, 340, 20, 70])  # [lat min, lat max, lon min, lon max]
ax.coastlines()
ax.gridlines()
plt.colorbar(
    im, ax=ax, ticks=levels, shrink=0.75, pad=0.04, label="Solar Radiation (J/m²)"
)

plt.savefig("outputs/02_solar_radiation_prediction.jpg")
```

**Comment by loliverhennigh** [comment]
02_solar_radiation_1h
02_solar_radiation_6h

Ok, so I after looking into this I think the results look decent. These images are for 1h and 6 hr model. The 6hr model accumulates the J/m2 over 6 hours and the 1 hr model for only 1 hour.

```
import os

os.makedirs("outputs", exist_ok=True)
from dotenv import load_dotenv

load_dotenv()  # TODO: make common example prep function

from earth2studio.data import GFS
from earth2studio.io import ZarrBackend
from earth2studio.models.dx import SolarRadiationAFNO1H
from earth2studio.models.px import SFNO

# Load the default model package which downloads the check point from NGC
package = SFNO.load_default_package()
prognostic_model = SFNO.load_model(package)

package = SolarRadiationAFNO1H.load_default_package()
diagnostic_model = SolarRadiationAFNO1H.load_model(package)

# Create the data source
data = GFS()

# Create the IO handler, store in memory
io = ZarrBackend()

import earth2studio.run as run

nsteps = 2
io = run.diagnostic(
    ["2021-06-01"], nsteps, prognostic_model, diagnostic_model, data, io
)

print(io.root.tree())

from datetime import datetime

import cartopy.crs as ccrs
import matplotlib.pyplot as plt
import numpy as np

forecast = datetime(2021, 6, 1)
variable = "ssrd"  # Surface solar radiation
step = 2  # lead time = 48 hrs

plt.close("all")

# Get the data range for levels
data_max = np.max(io[variable][0, step])
levels = np.linspace(0, data_max, 11)  # Range for solar radiation in J/m^2

# Define different viewing angles
viewpoints = [
    {"lon": -100, "lat": 40, "name": "north_america"},
    {"lon": 0, "lat": 0, "name": "africa"},
    {"lon": 120, "lat": 30, "name": "asia"},
    {"lon": -60, "lat": -30, "name": "south_america"},
]

fig, axes = plt.subplots(2, 2, figsize=(20, 12), subplot_kw={"projection": ccrs.PlateCarree()})
axes = axes.flatten()

for ax, viewpoint in zip(axes, viewpoints):
    # Set the projection for this axis
    projection = ccrs.Orthographic(viewpoint["lon"], viewpoint["lat"])
    ax = plt.subplot(2, 2, axes.tolist().index(ax) + 1, projection=projection)
    # Plot the field using contourf
    im = ax.contourf(
        io["lon"][:],
        io["lat"][:],
        io[variable][0, step],
        levels,
        transform=ccrs.PlateCarree(),
        cmap="YlOrRd",
    )
    ax.set_title(f"{forecast.strftime('%Y-%m-%d')} - {viewpoint['name'].title()} View\nLead time: {6*step}hrs")
    ax.coastlines()
    ax.gridlines()

# Add a single colorbar for all plots
cbar_ax = fig.add_axes([0.92, 0.15, 0.02, 0.7])
cbar = plt.colorbar(im, cax=cbar_ax, ticks=levels)
cbar.set_label("Solar Radiation (J/m²)")

plt.tight_layout(rect=[0, 0, 0.9, 1])
plt.savefig("outputs/02_solar_radiation_prediction_global.jpg", bbox_inches='tight', dpi=300)

```

File: earth2studio/models/dx/solarradiation_afno.py (hunk 2 above)
Comment (loliverhennigh) [reply earth2studio/models/dx/solarradiation_afno.py]: Hey  @albertocarpentieri, can you take a look at the doc strings for these models now and make sure they are right.

File: earth2studio/models/dx/solarradiation_afno.py (hunk 3 above)
Comment (loliverhennigh) [reply earth2studio/models/dx/solarradiation_afno.py]: I have a PR to get the model releases now. Ill update this when its ready and test again.

File: test/models/dx/test_solarradiation_afno.py (hunk 8 above)
Comment (loliverhennigh) [reply test/models/dx/test_solarradiation_afno.py]: Fixed

File: earth2studio/models/dx/solarradiation_afno.py (hunk 10 above)
Comment (loliverhennigh) [reply earth2studio/models/dx/solarradiation_afno.py]: Fixed

File: earth2studio/models/dx/solarradiation_afno.py (hunk 6 above)
Comment (loliverhennigh) [reply earth2studio/models/dx/solarradiation_afno.py:283]: Oh dang, yeah this is a bug

File: earth2studio/models/dx/solarradiation_afno.py (hunk 7 above)
Comment (loliverhennigh) [reply earth2studio/models/dx/solarradiation_afno.py]: So I rewrote this to run in a loop similar to the precipitation_afno_v2. The issue you raised here would have caused issues when running on multiple time or lead times. Its fixed now though.

**Comment by loliverhennigh** [comment]
Ok, I think everything is all addressed now (@NickGeneva). There are two things we need, 

1.  @albertocarpentieri, Can you take a look at this PR and sign off on the results if they look reasonable. If you have anything to validate against that would be fantastic as well.
2. The model is hopefully working its way through the publishing process and when its done I will update the ngc link.

File: earth2studio/models/dx/solarradiation_afno.py (hunk 6 above)
Comment (NickGeneva) [reply earth2studio/models/dx/solarradiation_afno.py:283]: The loop solution looks fine to me Oliver, not the best for perf but I think fine since these are usually getting pipelines anyway.

But can you add either a commend in the doc string of this function or preferable a check to make sure time and lead_time are of size 1 for future proofing to make sure we dont forget this limit.

**Review by NickGeneva — APPROVED** [review]

File: earth2studio/models/dx/solarradiation_afno.py (hunk 2 above)
Comment (albertocarpentieri) [reply earth2studio/models/dx/solarradiation_afno.py]: @loliverhennigh I would call the variable Surface Solar Radiation Downwards (SSRD) as in ERA5. Why did you remove the southpole?

Comment (NickGeneva) [reply earth2studio/models/dx/solarradiation_afno.py]: Resolved
